src/components/data-table.tsx

```diff
@@ -0,0 +1,239 @@
+/* eslint-disable react/require-default-props */
+import { ReactNode, HTMLAttributes } from 'react';
+import cn from 'classnames';
+
+import withDataLoader, { WrapperProps } from './data-loader';
+
+import '../styles/components/data-table.scss';
+
+type CommonColumn<T> = {
+  label?: ReactNode;
+  name: string;
+  // same function signature as a map function
+  render: (datum: T, index: number, data: T[]) => ReactNode;
+  width?: string;
+};
+
+// Either a column is sortable
+export interface SortableColumn<T> extends CommonColumn<T> {
+  sortable?: true;
+  sorted?: 'ascend' | 'descend';
+}
+// Or it's not sortable
+export interface NonSortableColumn<T> extends CommonColumn<T> {
+  sortable?: false | undefined;
+  sorted?: never;
+}
+
+type SharedProps<T> = {
+  /**
+   * An array of objects which specifies attributes about each column of your data. Each object has
+   * label, name and render attributes.
+   */
+  columns: Array<SortableColumn<T> | NonSortableColumn<T>>;
+  /**
+   * Table fixed layout
+   */
+  // Note sure why it doesn't detect it as used...
+  // eslint-disable-next-line react/no-unused-prop-types
+  fixedLayout?: boolean;
+};
+
+const BLOCK = 'data-table';
```

Comment:
Tried this approach here of putting the BEM's `Block` bit inside a variable reused throughout the file.
Not really necessary, but I thought it might be easier to follow BEM concepts that way, and would avoid typos in class names as those can't be caught by tools.